azurelinuxagent/ga/rsm_version_updater.py

```diff
@@ -0,0 +1,159 @@
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if
+                os.path.isdir(agent_dir)]
+
+    def _get_available_agents_on_disk(self):
+        available_agents = [agent for agent in self._get_all_agents_on_disk() if agent.is_available]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    def is_update_allowed_this_time(self):
+        """
+        update is allowed once per (as specified in the conf.get_autoupdate_frequency())
+        If update allowed, we update the last_attempted_rsm_version_update_time to current time.
+        """
+        now = datetime.datetime.now()
+
+        if self._last_attempted_rsm_version_update_time != datetime.datetime.min:
+            next_attempt_time = self._last_attempted_rsm_version_update_time + datetime.timedelta(
+                seconds=conf.get_agentupdate_frequency())
+        else:
+            next_attempt_time = now
+
+        if next_attempt_time > now:
+            return False
+        self._last_attempted_rsm_version_update_time = now
+        # The time limit elapsed for us to allow updates.
+        return True
+
+    def check_and_switch_updater_if_changed(self, agent_family, gs_id):
+        """
+        Checks if there is a new goal state and decide if we need to continue with rsm update or switch to self-update.
+        Firstly it checks agent supports GA versioning or not. If not, we raise exception to switch to self-update.
+        if vm is enabled for RSM updates and continue with rsm update, otherwise we raise exception to switch to self-update.
+        if either isVersionFromRSM or isVMEnabledForRSMUpgrades is missing in the goal state, we ignore the update as we consider it as invalid goal state.
+        """
+        if self._gs_id != gs_id:
+            self._gs_id = gs_id
+            if not conf.get_enable_ga_versioning():
+                raise VMDisabledRSMUpdates()
+
+            if agent_family.is_vm_enabled_for_rsm_upgrades is None:
+                raise AgentUpdateError(
+                    "Received invalid goal state:{0}, missing isVMEnabledForRSMUpgrades property. So, skipping agent update".format(
+                        gs_id))
+            elif not agent_family.is_vm_enabled_for_rsm_upgrades:
+                raise VMDisabledRSMUpdates()
+            else:
+                if agent_family.is_version_from_rsm is None:
+                    raise AgentUpdateError(
+                        "Received invalid goal state:{0}, missing isVersionFromRSM property. So, skipping agent update".format(
+                            gs_id))
+
+    def retrieve_agent_version(self, agent_family, goal_state):
+        """
+        Get the agent version from the goal state
+        """
+        if agent_family.version is None and agent_family.is_vm_enabled_for_rsm_upgrades and agent_family.is_version_from_rsm:
```

Comment:
We talked about encapsulating the use of those tri-state variables in 1 single method, otherwise they are difficult to use by callers (they always need to check for None). The method above, check_and_switch_updater_if_changed, is protecting itself against None, but this method is not.

Reply:
Yes, I felt version check None in check_and_switch_updater_if_changed method is not a good place. That's why I added in retrieve version method. 

Still adding confusion, I can place it in same method

Reply:
My suggestion is to encapsulate this in the GoalState class. Make the tri-state variables private and provide access to them only thru 1 method, let's say get_agent_update_info, which returns a struct with the two booleans and the string, and throws exceptions if the values in the goal state are invalid or missing.

Of course, there are other ways to do this. The idea is to limit the potential use of invalid values to 1 single method instead of having every single method that uses those properties checking for None.